### PR TITLE
allow running buildah on newer nodes

### PR DESCRIPTION
**Pull request by ibotty**
# Changes
This will allow running buildah on nodes with kernel 5.12 or later.

Buildah requires the `SETFCAP` capability to work.  This pull requests updates  OpenShift SCC to allow adding this capability and the included `buildah` clustertask to request it.

I do not have access to clusters with older kernels so I don't know whether it will work on these nodes.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

## Release Notes

```release-note
buildah ClusterTask is now allowed SETFCAP capability
```

### CODE REVIEW

**Comment by ibotty** [comment]
> @ibotty: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

No release notes are necessary (afaict), so what should I do?

**Comment by nikhil-thomas** [comment]
@gabemontero what are your thoughts 🧑‍💻

**Comment by nikhil-thomas** [comment]
> > @ibotty: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.
> 
> No release notes are necessary (afaict), so what should I do?

i think a 1 line release will help our users to be aware of this change.

but to get rid of the block without adding a release-note add 
```
```release-note
NONE
\```

```

at the end of our pr description.

**Comment by concaf** [comment]
```release-note
buildah ClusterTask is now allowed SETFCAP capability
```

**Comment by ibotty** [comment]
I added the release notes as suggested.  You are certainly right.

BTW: I plan to do a pull request in tektoncd's catalog when the review here is done.

**Comment by nikhil-thomas** [comment]
@piyush-garg do we need to add similar securityContext blocks to all the s2i tasks in our addons [here](https://github.com/tektoncd/operator/tree/main/cmd/openshift/operator/kodata/tekton-addon/addons/02-clustertasks/source_local)

**Comment by gabemontero** [comment]
note: as I understand it, "the problem" and need for this new permission only applies to running on OCP 4.11

**Comment by nikhil-thomas** [comment]
/hold

let us get some more 👀 on this.
🧑‍💻

**Comment by piyush-garg** [comment]
> 

It did not work for me for a normal user, though worked for kubeadmin on ocp 4.11

```
task build has failed: failed to create task run pod "s2i-go-1167-ubi8-run-f5ms6-build": pods "s2i-go-1167-ubi8-run-f5ms6-build-pod" is forbidden: unable to validate against any security context constraint: [provider "anyuid": Forbidden: not usable by user or serviceaccount, spec.containers[1].securityContext.capabilities.add: Invalid value: "CAP_SETFCAP": capability may not be added, spec.containers[2].securityContext.capabilities.add: Invalid value: "CAP_SETFCAP": capability may not be added, provider "restricted": Forbidden: not usable by user or serviceaccount, provider "nonroot-v2": Forbidden: not usable by user or serviceaccount, provider "hostmount-anyuid": Forbidden: not usable by user or serviceaccount, provider "machine-api-termination-handler": Forbidden: not usable by user or serviceaccount, provider "hostnetwork-v2": Forbidden: not usable by user or serviceaccount, provider "hostnetwork": Forbidden: not usable by user or serviceaccount, provider "hostaccess": Forbidden: not usable by user or serviceaccount, provider "node-exporter": Forbidden: not usable by user or serviceaccount, provider "privileged": Forbidden: not usable by user or serviceaccount]. Maybe missing or invalid Task test/s2i-go
```

**Comment by piyush-garg** [comment]
> @ibotty is this PR tested on any OpenShift cluster? I am not able to run the tasks as a normal user

It did not work for me as a normal user, but worked as a kubeadmin

**Comment by ibotty** [comment]
@piyush-garg: Are you sure you the `pipelines-scc` got updated?  Mine looks like

```
oc get scc pipelines-scc
NAME            PRIV    CAPS          SELINUX     RUNASUSER   FSGROUP     SUPGROUP   PRIORITY   READONLYROOTFS   VOLUMES
pipelines-scc   false   ["SETFCAP"]   MustRunAs   RunAsAny    MustRunAs   RunAsAny   10         false            ["configMap","downwardAPI","emptyDir","persistentVolumeClaim","projected","secret"]
```

**Comment by nalind** [comment]
> for @nikhil-thomas 's question to myself, @adambkaplan, and @sbose78 about the security implications, I'm not sure I feel qualified to respond with confidence on how good or bad that additional change is, but I believe @nalind did tell me this was not a temporary state of affairs. That it was what would be needed moving forward.

That's right, I have no reason to expect the kernel to revert this.

> @nalind - do you have any input on the relative safety or lack there of around CAP_SETFCAP

I'm not an expert on that.

>     * lastly, when @nalind and I last talked, he mentioned he had buildah changes he needed to make to get things to work even with CAP_SETFCAP.  @nalind - were you in fact able to make those changes yet?  @nikhil-thomas said they are now using `registry.redhat.io/rhel8/buildah@sha256:23fb7971ea6ac4aaaaa1139473a602df0df19222a3b5a76b551b2b9ddd92e927` and that it is working.  Just wanted to make sure that lines up with your expectations.

I must have misspoken.  The most we can do there is check for that capability (near the existing check for CAP_SYS_ADMIN) and suggest it as a possible cause when we fail to set up the ID mappings.

**Comment by ibotty** [comment]
Argh!  I forgot to mention,  I am using the task with `BUILDER_IMAGE=quay.io/buildah/stable:latest`, because of missing redhat registry credentials.

**Comment by adambkaplan** [comment]
It should be noted that `CAP_SETFCAP` is permitted under the Baseline pod security standard:

https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline

**Comment by gabemontero** [comment]
> It should be noted that `CAP_SETFCAP` is permitted under the Baseline pod security standard:
> 
> https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline

good catch @adambkaplan forgot to look there

yes that is probably the best answer to say it is "OK" for `pipeline` SA to pick this up

**Comment by piyush-garg** [comment]
> @piyush-garg: Are you sure you the `pipelines-scc` got updated? Mine looks like
> 
> ```
> oc get scc pipelines-scc
> NAME            PRIV    CAPS          SELINUX     RUNASUSER   FSGROUP     SUPGROUP   PRIORITY   READONLYROOTFS   VOLUMES
> pipelines-scc   false   ["SETFCAP"]   MustRunAs   RunAsAny    MustRunAs   RunAsAny   10         false            ["configMap","downwardAPI","emptyDir","persistentVolumeClaim","projected","secret"]
> ```

Yes, i see it like this and not able to create pod. taskrun not able to convert to pod

**Comment by piyush-garg** [comment]
@ibotty This PR https://github.com/tektoncd/operator/pull/813 worked for me

**Comment by sm43** [comment]
I tried this pr on OCP 4.11
by creating a non admin user 
got error as below
```
failed to create task run pod "buildah-run-kqc8j-run-buildah":
              pods "buildah-run-kqc8j-run-buildah-pod" is forbidden: unable to validate
              against any security context constraint: [provider "anyuid": Forbidden:
              not usable by user or serviceaccount, spec.containers[0].securityContext.capabilities.add:
              Invalid value: "CAP_SETFCAP": capability may not be added, spec.containers[1].securityContext.capabilities.add:
              Invalid value: "CAP_SETFCAP": capability may not be added, spec.containers[2].securityContext.capabilities.add:
              Invalid value: "CAP_SETFCAP": capability may not be added, provider
              "restricted": Forbidden: not usable by user or serviceaccount, provider
              "nonroot-v2": Forbidden: not usable by user or serviceaccount, provider
              "hostmount-anyuid": Forbidden: not usable by user or serviceaccount,
              provider "machine-api-termination-handler": Forbidden: not usable by
              user or serviceaccount, provider "hostnetwork-v2": Forbidden: not usable
              by user or serviceaccount, provider "hostnetwork": Forbidden: not usable
              by user or serviceaccount, provider "hostaccess": Forbidden: not usable
              by user or serviceaccount, provider "node-exporter": Forbidden: not
              usable by user or serviceaccount, provider "privileged": Forbidden:
              not usable by user or serviceaccount]. Maybe missing or invalid Task
              test/buildah
```

steps performed
- OCP 4.11
- Created a user following https://docs.openshift.com/container-platform/4.10/authentication/identity_providers/configuring-htpasswd-identity-provider.html#identity-provider-overview_configuring-htpasswd-identity-provider
-  Changed user to non admin one
- Created a Pipelinerun using buildah

**Comment by ibotty** [comment]
Argh.  While trying to debug the problem I noticed, that I did make a copy-paste error:  the capabilities are not prepended with `CAP_` in kubernetes.  My "fixed" task had `addCapabilities: [SET_FCAP]` set.
:facepalm: 

Should I update the PR to add (the right capability) to all other s2i tasks as well?

File: cmd/openshift/operator/kodata/tekton-addon/addons/02-clustertasks/source_local/buildah/buildah-task.yaml [hunk 1]
```diff
@@ -78,10 +81,16 @@ spec:
...
     - name: varlibcontainers
       mountPath: /var/lib/containers
+    securityContext:
+      capabilities:
+        add: ["SETFCAP"]
 
   - name: digest-to-results
     image: $(params.BUILDER_IMAGE)
     script: cat $(workspaces.source.path)/image-digest | tee /tekton/results/IMAGE_DIGEST
+    securityContext:
```
Comment (piyush-garg) [inline cmd/openshift/operator/kodata/tekton-addon/addons/02-clustertasks/source_local/buildah/buildah-task.yaml:91]: We dont need to the capability for this step

**Review by vdemeester — APPROVED** [review]

**Comment by piyush-garg** [comment]
> Should I update the PR to add (the right) capability to all other s2i tasks as well?

Yes we need to do the change for all s2i tasks too, but only for the build and push steps, not all

File: cmd/openshift/operator/kodata/tekton-addon/addons/02-clustertasks/source_local/s2i-dotnet/s2i-dotnet-task.yaml [hunk 2]
```diff
@@ -59,13 +59,19 @@ spec:
           mountPath: /var/lib/containers
         - name: gen-source
           mountPath: /gen-source
+      securityContext:
+          capabilities:
```
Comment (piyush-garg) [inline cmd/openshift/operator/kodata/tekton-addon/addons/02-clustertasks/source_local/s2i-dotnet/s2i-dotnet-task.yaml:63]: indentation is wrong here

Comment (ibotty) [reply cmd/openshift/operator/kodata/tekton-addon/addons/02-clustertasks/source_local/s2i-dotnet/s2i-dotnet-task.yaml]: Sorry I missed that.  Fixed.

**Comment by sbose78** [comment]
Followed up, looks good 👍

**Comment by vdemeester** [comment]
/lgtm
I think we can lift the `hold`

**Pull request merged**